packages/salesforcedx-apex-debugger/src/adapter/apexDebug.ts

```diff
@@ -59,15 +71,376 @@ import {
 import { nls } from '../messages';
 import os = require('os');
 
+const TRACE_ALL = 'all';
+const TRACE_CATEGORY_VARIABLES = 'variables';
+const TRACE_CATEGORY_LAUNCH = 'launch';
+const TRACE_CATEGORY_PROTOCOL = 'protocol';
+
+export type TraceCategory = 'all' | 'variables' | 'launch' | 'protocol';
+
 export interface LaunchRequestArguments
   extends DebugProtocol.LaunchRequestArguments {
+  /** comma separated list of trace selectors (see TraceCategory)
+	  */
+  trace?: boolean | string;
   userIdFilter?: string;
   requestTypeFilter?: string;
   entryPointFilter?: string;
   sfdxProject: string;
 }
 
-export class ApexDebug extends DebugSession {
+export class ApexDebugStackFrameInfo {
+  public readonly requestId: string;
+  public readonly frameNumber: number;
+  public globals: Value[];
+  public statics: Value[];
+  public locals: LocalValue[];
+  public references: Reference[];
+  constructor(requestId: string, frameNumber: number) {
+    this.requestId = requestId;
+    this.frameNumber = frameNumber;
+  }
+}
+
+export enum ApexVariableKind {
+  Global = 10,
+  Static = 20,
+  Local = 30,
+  Field = 40,
+  Collection = 50
+}
+
+export class ApexVariable extends Variable {
+  public readonly declaredTypeRef: string;
+  private readonly slot: number;
+  private readonly kind: ApexVariableKind;
+
+  constructor(
+    value: Value,
+    kind: ApexVariableKind,
+    variableReference?: number
+  ) {
+    super(value.name, ApexVariable.valueAsString(value), variableReference);
+    this.declaredTypeRef = value.declaredTypeRef;
+    this.kind = kind;
+    if ((value as LocalValue).slot !== undefined) {
+      this.slot = (value as LocalValue).slot;
+    } else {
+      this.slot = Number.MAX_SAFE_INTEGER;
+    }
+  }
+
+  public static valueAsString(value: Value): string {
+    if (typeof value.value === 'undefined' || value.value === null) {
+      // We want to explicitly display null for null values (no type info for strings).
+      return ApexVariable.isString(value)
+        ? 'null'
+        : `null [${value.nameForMessages}]`;
+    }
+
+    if (ApexVariable.isString(value)) {
+      // We want to explicitly quote string values like in Java. This allows us to differentiate null from 'null'.
+      return `'${value.value}'`;
+    }
+
+    return `${value.value} [${value.nameForMessages}]`;
```

Comment:
VS Code debugger UI gets crowded pretty quickly because we can't move them away from the left hand side. I'm more in favor of showing the type using https://github.com/Microsoft/vscode-debugadapter-node/blob/master/protocol/src/debugProtocol.ts#L1260.

Reply:
Done.